### PR TITLE
Variable lacunarity node

### PR DESCRIPTION
**Pull request by kalwalt**
As discussed time ago in this [issue](https://github.com/nortikin/sverchok/issues/1205) i propose this node, I added the seed option as the turbulence node does, (maybe we can store that funcs outside in a dedicated file?) because i could add the seed option also to the fractal node that lacks this feature.
I think we can make some optimization to the code, i'm waiting for a review, @zeffii.

### CODE REVIEW

**Comment by zeffii** [comment]
yeah, make the seed injection a library func.. in utils.. perhaps   `/utils/seed_funcs.py`

**Comment by zeffii** [comment]
i find `SvVariableNode` a weak name, how about `SvLacunarityNode` - Variable is uninformative.

**Comment by kalwalt** [comment]
> i find SvVariableNode a weak name, how about SvLacunarityNode - Variable is uninformative.

you are right, i have no `fantasy` for the names, i will change it. SvLacunarityNode is much better.

**Comment by kalwalt** [comment]
> yeah, make the seed injection a library func.. in utils.. perhaps /utils/seed_funcs.py

Making also the changes in the `Fractal` and `Turbulence` Node.

**Comment by kalwalt** [comment]
i missed to change the class name, will fix it.

**Comment by kalwalt** [comment]
image example:

variable_lacu_example

File: utils/sv_seed_funcs.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import bpy
```
Comment (kalwalt) [inline utils/sv_seed_funcs.py]: import bpy is needed?

Comment (zeffii) [inline utils/sv_seed_funcs.py]: nope.. not if it isn't used directly.

**Comment by kalwalt** [comment]
with https://github.com/nortikin/sverchok/pull/1425/commits/1f1cf8b3e05fec613fbc396cc7bbcc8c646830c2 i get this error, when i try to switch from fractal to other multi_fractal, and other: 
```python
File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/core/update_system.py", line 322, in do_update_general
    node.process()
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 151, in process
    _seed = inputs['Seed'].sv_get()[0][0]
Node Vector Fractal had exception 'bpy_prop_collection[key]: key "Seed" not found'

```
This error is due to the fact that Seed is not in the wrapped_update func?

**Comment by kalwalt** [comment]
This solved:
```diff
- seed = IntProperty(default=0, name='Seed', update=updateNode)
+ seed = IntProperty(default=0, name='Seed', update=wrapped_update)
```

**Comment by zeffii** [comment]
i don't see why you need wrapped_update as the update function for seed../!?

**Comment by kalwalt** [comment]
> i don't see why you need wrapped_update as the update function for seed../!?

i will try again but seems that is needed...

**Comment by zeffii** [comment]
I also don't see it immediately.. but we don't look ghosts, so let's look closer :)

**Comment by kalwalt** [comment]
it's not ok. i have now in Ridged, Hybrid, and Hetero option a double Seed socket. should be fixed in other way.

**Comment by zeffii** [comment]
lets reason about it. 

setting the seed should only influence the offset of the vectors (if the function doesn't support seed) or just allow you to pass seed when it does.

**Comment by kalwalt** [comment]
maybe i need to change the num inputs in the wrapped function? since i add one more it can not  found it.

**Comment by zeffii** [comment]
yep. i think so too, after adding the Seed socket that number is off by one or something

**Comment by zeffii** [comment]
if the seed is always available, don't stick it in as the last socket.

**Comment by kalwalt** [comment]
> if the seed is always available, don't stick it in as the last socket.

ok i will move it in a upper position.

**Comment by zeffii** [comment]
the other day we added a socket property called `hide_safe`  

sv_init could be 
```
    def wrapped_update(self, context):
        num_inputs = len(self.inputs)
        enabled = props_enabled.get(self.fractal_type)

        if enabled == (0, 0):
            ...

        elif enabled == (1, 0):
            ...

        elif enabled == (1, 1):
            self.inputs['Offset'].hide_safe = False
            self.inputs['Gain'].hide_safe = False

        

    def sv_init(self, context):
        self.inputs.new('VerticesSocket', 'Vertices')
        self.inputs.new('StringsSocket', 'H Factor').prop_name = 'h_factor'
        self.inputs.new('StringsSocket', 'Lacunarity').prop_name = 'lacunarity'
        self.inputs.new('StringsSocket', 'Octaves').prop_name = 'octaves'
        self.inputs.new('StringsSocket', 'Seed').prop_name = 'seed'
        self.inputs.new('StringsSocket', 'Offset').prop_name = 'offset'
        self.inputs.new('StringsSocket', 'Gain').prop_name = 'gain'
        self.outputs.new('StringsSocket', 'Value')
        self.wrapped_update(context)

**Comment by zeffii** [comment]
first add all sockets in the sv_init, then selectively `hide_safe = True/False` in the wrapped_update whenever the mode changes.

**Comment by kalwalt** [comment]
I will do tomorrow, the `hide_safe` changes. Seems that was the number of inputs the cause of the issue.

**Comment by zeffii** [comment]
If you get it working without `hide_safe`, that's great too. because i'm not certain hidden sockets are supported by Monads yet.

**Comment by zeffii** [comment]
i think the dynamic sockets (removing, recreating) is about the least coder friendly part of Sverchok, to some extent it is related to the barebones PyNodes api... but it's also definitely our fault for accepting the more verbose way for so long..

**Comment by kalwalt** [comment]
There is this issue with fractal node (not only this branch but also the master is affected): while switching from one type to another the type will ot change from

```
FRACTAL
to
MULTI_FRACTAL
```

```
RIDGED_MULTI_FRACTAL
to
HYBRID_MULTI_FRACTAL  
```
in def process add a print like this and you will see that  
inputs, outputs = self.inputs, self.outputs
print('fractal_type is: ', self.fractal_type)

**Comment by zeffii** [comment]
ok. let psuedo
```
sockets_per_mode = {
    fractal:       verts, seed, h_factor, lacunarity, octaves
    multifractal:  verts, seed, h_factor, lacunarity, octaves
    hetero:        verts, seed, h_factor, lacunarity, octaves, offset
    ridged:        verts, seed, h_factor, lacunarity, octaves, offset, gain
    hybrid:        verts, seed, h_factor, lacunarity, octaves, offset, gain
}
```

**Comment by zeffii** [comment]
come up with a wrapped_update function that correctly removes / keeps the right sockets.. i suggest starting from scratch.

go completely braindead 
```
# states with the same sockets
'A' : FRACTAL, MULTI_FRACTAL
'B' : HETERO 
'C' : RIDGED, HYBRID

# state switches
if A to B:
    add octaves
if B to A:
    remove octaves
if B to C:
    add gain
if C to B:
    remove gain
if C to A:
    remove offset, gain
if A to C:
    add offset, gain
```

**Comment by zeffii** [comment]
now you can track the current state using a variable... or do a lookup on the current sockets, before changing them.   `if offset and gain in sockets, then i'm current mode C `

**Comment by zeffii** [comment]
overall that's probably nice and short code :) and easier to comprehend that the current slightly painful code : )

**Comment by kalwalt** [comment]
Ok i think i understood the concept, i'm working on it, eventually i will ask some other question about.

**Comment by zeffii** [comment]
something like this?   partial
```python
current_mode = {5: 'A', 6: 'B', 7: 'C'}.get(len(node.inputs))
new_mode = self.fractal_type
modes = {
    'A' : ('FRACTAL', 'MULTI_FRACTAL'),
    'B' : ('HETERO',), 
    'C' : ('RIDGED', 'HYBRID')
}
change = (current_mode, new_mode)
# state switches
if change == ('A', 'B'):
    add octaves
if change == ('B', 'A'):
    remove octaves
if B to C:
    add gain
if C to B:
    remove gain
if C to A:
    remove offset, gain
if A to C:
    add offset, gain

```

**Comment by zeffii** [comment]
maybe succinctly
```python
current_mode = {5: 'A', 6: 'B', 7: 'C'}.get(len(self.inputs))
new_mode = {
    'FRACTAL': 'A', 'MULTI_FRACTAL': 'A' 'HETERO': 'B', 'RIDGED': 'C', 'HYBRID': 'C'
}.get(self.fractal_type)

change = (current_mode, new_mode)

# state switches
if change == ('A', 'B'):
    add octaves
if change == ('B', 'A'):
    remove octaves
if change == ('B', 'C'):
    add gain
if change == ('C', 'B'):
    remove gain
if change == ('C', 'A'):
    remove offset, gain
if change == ('A', 'C'):
    add offset, gain
```

**Comment by kalwalt** [comment]
very efficient!
but to add a socket i should:
```python
# adding octaves
 self.inputs.new('StringsSocket', 'Octaves').prop_name = 'octaves'
```
thers is only this way?

**Comment by zeffii** [comment]
maybe less chaos
```python
# out of class
socket_count_to_mode = {5: 'A', 6: 'B', 7: 'C'}
fractal_type_to_mode = {'FRACTAL': 'A', 'MULTI_FRACTAL': 'A' 'HETERO': 'B', 'RIDGED': 'C', 'HYBRID': 'C'}

...
current_mode = socket_count_to_mode.get(len(self.inputs))
new_mode = fractal_type_to_mode.get(self.fractal_type)

change = (current_mode, new_mode)

# state switches
if change == ('A', 'B'):
    add octaves
if change == ('B', 'A'):
    remove octaves
if change == ('B', 'C'):
    add gain
if change == ('C', 'B'):
    remove gain
if change == ('C', 'A'):
    remove offset, gain
if change == ('A', 'C'):
    add offset, gain
```

**Comment by zeffii** [comment]
> 
``` # adding octaves
 self.inputs.new('StringsSocket', 'Octaves').prop_name = 'octaves'
```

yeah, i think this is one time where it might help to be more verbose in that if / elif  chain..

**Comment by zeffii** [comment]
else you get some serious voodoo like
```python
# out of class
socket_count_to_mode = {5: 'A', 6: 'B', 7: 'C'}
fractal_type_to_mode = {'FRACTAL': 'A', 'MULTI_FRACTAL': 'A' 'HETERO': 'B', 'RIDGED': 'C', 'HYBRID': 'C'}

    # class function
    def modify_input_sockets(self, action, *sockets):
        modify = getattr(self.inputs, action)
        for socket in sockets:
            if action is 'remove':
                modify(self.inputs[socket.title()])
            elif action is 'new':
                modify(socket.title()).prop_name = socket

    def wrapped_update(....):
        current_mode = socket_count_to_mode.get(len(self.inputs))
        new_mode = fractal_type_to_mode.get(self.fractal_type)

        change = (current_mode, new_mode)

        # state switches
        if change == ('A', 'B'):
            self.modify_input_sockets('new', 'octaves')
        if change == ('B', 'A'):
            self.modify_input_sockets('remove', 'octaves')
        if change == ('B', 'C'):
            new gain   ..etc
        if change == ('C', 'B'):
            remove gain
        if change == ('C', 'A'):
            remove offset, gain
        if change == ('A', 'C'):
            new offset, gain
```

**Comment by zeffii** [comment]
```python
    # class functions
    def rm_input_sockets(self, *sockets):
        for socket in sockets:
            self.inputs.remove(self.inputs[socket.title()])

    def mk_input_sockets(self, *sockets):
        for socket in sockets:
            self.inputs.new(socket.title()).prop_name = socket



...
self.mk_input_sockets('octaves')

...
self.rm_input_sockets('offset', 'gain')
```

**Comment by kalwalt** [comment]
But the fractal switch doesn't works, don't know why.

**Comment by zeffii** [comment]
fractal_type_to_mode   doesn't have the full string names..

**Comment by kalwalt** [comment]
From the latest:
```
Traceback (most recent call last):
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 114, in wrapped_update
    self.mk_input_sockets('offset')
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 97, in mk_input_sockets
    self.inputs.new(socket.title()).prop_name = socket
TypeError: NodeInputs.new(): required parameter "name" not specified
File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 103, in wrapped_update
```
i think it's related to the name of the sockets, but changing it from 'offset' to Offset not help, there is something else?

**Comment by zeffii** [comment]
then 

    self.inputs.new(socket.title()).prop_name = socket

needs to be

    self.inputs.new('StringSocket', socket.title()).prop_name = socket

**Comment by zeffii** [comment]
```python
        if change == ('C', 'A'):
            # remove offset, gain
            self.rm_input_sockets('offset', 'gain')
        if change == ('A', 'C'):
            # add offset, gain
            self.mk_input_sockets('offset', 'gain')

```

**Comment by zeffii** [comment]
be sure to make it an if/elif/elif/elif ..etc
```python
        elif change == ('C', 'A'):
            # remove offset, gain
            self.rm_input_sockets('offset', 'gain')
        elif change == ('A', 'C'):
            # add offset, gain
            self.mk_input_sockets('offset', 'gain')

```

**Comment by kalwalt** [comment]
ok thank you but with:
`self.inputs.new('StringSocket', socket.title()).prop_name = socket
`
i get:
```
HETERO_TERRAIN
Offset
Traceback (most recent call last):
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 117, in wrapped_update
    self.mk_input_sockets('offset')
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 100, in mk_input_sockets
    self.inputs.new('StringSocket', socket.title()).prop_name = socket
AttributeError: 'NodeSocket' object has no attribute 'prop_name'
File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 106, in wrapped_update


```

**Comment by zeffii** [comment]
`StringsSocket`

**Comment by zeffii** [comment]
Github Issue tracker doesn't notify readers when previous posts have been updated...  it really is one feature that would save much brainfuckery

**Comment by kalwalt** [comment]
oh yes! i was blinded...
```diff
- StringSocket
+ StringsSocket
```

> Github Issue tracker doesn't notify readers when previous posts have been updated... it really is one feature that would save much brainfuckery

it will save a lot...

**Comment by kalwalt** [comment]
Anyway not solve the switch issue, now i get the correct sockets in the node but still doesn't switch `from FRACTAL to MULTI_FRACTAL `  as before. it should works but not.

**Comment by zeffii** [comment]
but don't fractal and multi fractal have the same sockets?

**Comment by zeffii** [comment]
and also this..
```
        elif change == ('C', 'A'):
            # remove offset, gain
            self.rm_input_sockets('offset', 'gain')
        elif change == ('A', 'C'):
            # add offset, gain
            self.rm_input_sockets('offset', 'gain')
```
spot the bug?

**Comment by zeffii** [comment]
if we can do this...
```python

        change = (current_mode, new_mode)
        # state switches
        if change == ('A', 'B'):
            # add offset
            self.mk_input_sockets('offset')
        elif change == ('B', 'A'):
            # remove offset
            self.rm_input_sockets('offset')
        elif change == ('B', 'C'):
            # add gain
            self.mk_input_sockets('gain')
        elif change == ('C', 'B'):
            # remove gain
            self.rm_input_sockets('gain')
        elif change == ('C', 'A'):
            # remove offset, gain
            self.rm_input_sockets('offset', 'gain')
        elif change == ('A', 'C'):
            # add offset, gain
            self.mk_input_sockets('offset', 'gain')
```

then we can do this too 

```python

        change = (current_mode, new_mode)
        actionables = {
            ('A', 'B'):  (self.mk_input_sockets, ('offset',)),
            ('B', 'A'):  (self.rm_input_sockets, ('offset',)),
            ('B', 'C'):  (self.mk_input_sockets, ('gain',)),
            ('C', 'B'):  (self.rm_input_sockets, ('gain',)),
            ('A', 'C'):  (self.mk_input_sockets, ('offset', 'gain')),
            ('C', 'A'):  (self.rm_input_sockets, ('offset', 'gain'))
        }.get(change)
        if actionables:
            socket_func, names = actionables
            socket_func(*names)

```

**Comment by kalwalt** [comment]
the same sockets but not the same function. The sockets remain the same (as should) but the function not: i mean the wrapped_fractal_func.

edit: oh i see your post now.

**Comment by zeffii** [comment]
yikes..  i should stop now.

```python

        change = (current_mode, new_mode)
        add = self.mk_input_sockets
        remove = self.rm_input_sockets
        actionables = {
            ('A', 'B'):  (add, ('offset',)),
            ('B', 'A'):  (remove, ('offset',)),
            ('B', 'C'):  (add, ('gain',)),
            ('C', 'B'):  (remove, ('gain',)),
            ('A', 'C'):  (add, ('offset', 'gain')),
            ('C', 'A'):  (remove, ('offset', 'gain'))
        }.get(change)
        if actionables:
            socket_func, names = actionables
            socket_func(*names)

```

**Comment by zeffii** [comment]
oh Noe!! suddenly a flash of Lisp there. :)

**Comment by zeffii** [comment]
something about this seems wrong..
```python
        out = []
        for idx, vlist in enumerate(verts):
            # lazy generation of full parameters.
            params = [(param[idx] if idx < len(param) else param[-1]) for param in param_list]
            final_vert_list = [seed_adjusted(vlist, _seed)]

        for fvlist in final_vert_list:
            out.append(wrapped_fractal_function(_noise_type, fvlist, *params))

```

**Comment by kalwalt** [comment]
yes maybe the parmeters?

**Comment by zeffii** [comment]
no.. indendation...
        out = []
        for idx, vlist in enumerate(verts):
            # lazy generation of full parameters.
            params = [(param[idx] if idx < len(param) else param[-1]) for param in param_list]
            final_vert_list = [seed_adjusted(vlist, _seed)]

            for fvlist in final_vert_list:
                out.append(wrapped_fractal_function(_noise_type, fvlist, *params))
``` 
?

**Comment by zeffii** [comment]
well maybe also the params... but definitely also the indendation.

**Comment by kalwalt** [comment]
ok no problem i will check it!

**Comment by kalwalt** [comment]
to be sure that the fractal_type is updated in the class i added:
`fractal_t = StringProperty()`
in def wrapped_update(self, context):
`fractal_t = self.fractal_type`
and finally in def process():
`wrapped_fractal_function = fractal_f[self.fractal_t]`
but i get : 
```
fractal_type is:  FRACTAL

  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/core/update_system.py", line 322, in do_update_general
    node.process()
  File "/home/walter/blender-2.78a-linux-glibc211-x86_64/2.78/scripts/addons/sverchok-master/nodes/vector/fractal.py", line 173, in process
    wrapped_fractal_function = fractal_f[self.fractal_t]
Node Vector Fractal had exception ''


```

**Comment by zeffii** [comment]
no .. EnumProperty  is a String already, (the enum you pick is the str value of that property when evaluated ) .. exception` '' `means  an empty key.

**Comment by zeffii** [comment]
additionally, the `wrapped_update` needs an explicit call to update at the end..  `updateNode(self, context)` i think..

**Comment by zeffii** [comment]
The params need proper vectorization.. if you plan to allow different parameter setting for each set of incoming vertices

**Comment by zeffii** [comment]
```
        out = []
        for idx, vlist in enumerate(verts):
            # lazy generation of full parameters.
            params = [(param[idx] if idx < len(param) else param[-1]) for param in param_list]
            final_vert_list = [seed_adjusted(vlist, _seed)]

            for fvlist in final_vert_list:
                out.append(wrapped_fractal_function(_noise_type, fvlist, *params))

        outputs[0].sv_set(out)
```
i should clarify that the term 'lazy generation of full parameters' doesn't mean that code is lazy, it refers to the style of not creating values until they're requested..

**Comment by zeffii** [comment]
uhmm...
```
        out = []
        for idx, vlist in enumerate(verts):
            # lazy generation of full parameters.
            params = [(param[idx] if idx < len(param) else param[-1]) for param in param_list]
            final_vert_list = [seed_adjusted(vlist, _seed)]
            out.append(wrapped_fractal_function(_noise_type, final_vert_list, *params))
```

**Comment by zeffii** [comment]
the `wrapped_fractal_function` already iterates over the incoming verts for you.

**Comment by kalwalt** [comment]
@zeffii with the latest seems that i fixed finally the issue, i need to do other checks and other small fixes i think, doing the docs for the new node and it will be fine.

**Comment by kalwalt** [comment]
Adding `updateNode(self, context)` to the wrapped_update was crucial i think....!

**Comment by kalwalt** [comment]
and also adding the `hide_safe` as mentioned before.

**Comment by zeffii** [comment]
don't use `hide_safe` if you can use   
```
        add = self.mk_input_sockets
        remove = self.rm_input_sockets
```

**Comment by kalwalt** [comment]
ok, thanks i will re-check tomorrow all the code, for now it'is enough.

File: nodes/vector/fractal.py [hunk 2]
```diff
@@ -70,10 +71,15 @@ def hybrid(nbasis, verts, h_factor, lacunarity, octaves, offset, gain):
     ('HYBRID_MULTI_FRACTAL', 4, (1, 1), hybrid),
 ]
 
+socket_count_to_mode = {5: 'A', 6: 'B', 7: 'C'}
+fractal_type_to_mode = {'FRACTAL': 'A',
```
Comment (zeffii) [inline nodes/vector/fractal.py]: this indentation is a little akward.

Comment (zeffii) [inline nodes/vector/fractal.py]: Especially the body of this dict

**Comment by zeffii** [comment]
better!

**Comment by zeffii** [comment]
less code _and_ more pleasant to scan through :)

**Comment by kalwalt** [comment]
some small things also to do to the `Variable lacunarity node`

**Comment by kalwalt** [comment]
In the `Variable lacunarity node` def draw_label does do anything? it is there because come from the `noise node` that i used as base.

**Comment by zeffii** [comment]
@kalwalt   look at your texture viz node's `draw_label func`

**Comment by zeffii** [comment]
now this is just nitpicking.. but you might agree
```python
    def wrapped_update(self, context):
        add = self.mk_input_sockets
        remove = self.rm_input_sockets

        current_mode = socket_count_to_mode.get(len(self.inputs))
        new_mode = fractal_type_to_mode.get(self.fractal_type)

        actionables = {
            ('A', 'B'): (add, ('offset',)),
            ('B', 'A'): (remove, ('offset',)),
            ('B', 'C'): (add, ('gain',)),
            ('C', 'B'): (remove, ('gain',)),
            ('A', 'C'): (add, ('offset', 'gain')),
            ('C', 'A'): (remove, ('offset', 'gain'))
        }.get((current_mode, new_mode))

        if actionables:
            socket_func, names = actionables
            socket_func(*names)

        updateNode(self, context)

```

**Comment by zeffii** [comment]
and even..
```python
        actionables = {
            'AB': (add, ('offset',)),
            'BA': (remove, ('offset',)),
            'BC': (add, ('gain',)),
            'CB': (remove, ('gain',)),
            'AC': (add, ('offset', 'gain')),
            'CA': (remove, ('offset', 'gain'))
        }.get(current_mode + new_mode)
```

**Comment by zeffii** [comment]
less symbols that might pose as an obfuscating factor to a new reader..

**Comment by kalwalt** [comment]
i vote for the last option, i'm going to try it.

**Comment by zeffii** [comment]
let me know when to merge.

**Comment by kalwalt** [comment]
just one thing more in the docs and will be ok.

**Comment by kalwalt** [comment]
ok i think is ready to merge.

**Pull request merged**